docs: mm2 handling large volumes of messages (#6636)
strimzi/strimzi-kafka-operator: documentation/modules/configuring/con-config-mirrormaker2-high-volume-messages.adoc

// Module included in the following assemblies:
//
// assembly-config-mirrormaker2.adoc

[id='con-mirrormaker-high-volume-messages-{context}']
= Handling high volumes of messages

[role="_abstract"]
If your MirrorMaker 2.0 deployment is going to be handling a high volume of messages, you might need to adjust its configuration to support it.

The flush pipeline for data replication is *source topic -> (Kafka Connect) source message queue -> producer buffer -> target topic*.
An offset flush timeout period (`offset.flush.timeout.ms`) is the time to wait for the producer buffer (`producer.buffer.memory`) to flush and offset data to be committed.

> Try to avoid a situation where a large producer buffer and an insufficient offset flush timeout period causes a _failed to flush_ type of error.

Reviewer: We actually have 1 common cause, but 2 different error messages: 
- "Failed to flush" is related to `producer.buffer.memory`. When this buffer is full, the producer's `send()` operation becomes synchronous and it slows down up to the point it hits timeout. 
- "Failed to commit offsets" is related to `offset.flush.timeout.ms`. KafkaConnect commits the offset in the source cluster periodically (every 1 minute by default), but it may hit timeout when there are too many messages during this interval.

> Try to avoid a situation where a large producer buffer and an insufficient offset flush timeout period causes a _failed to flush_ type of error.

Reviewer: Failed to flush/Failed to commit offsets-

* Increasing the number of nodes for the workers that run tasks using the `replicas` property

.Example MirrorMaker 2.0 configuration for handling high volumes of messages
[source,yaml,subs="+quotes,attributes"]
----
apiVersion: {KafkaMirrorMaker2ApiVersion}
kind: KafkaMirrorMaker2
metadata:
  name: my-mirror-maker2
spec:
  version: {DefaultKafkaVersion}
  replicas: 5
  connectCluster: "my-cluster-target"
  clusters:
  - alias: "my-cluster-source"
    bootstrapServers: my-cluster-source-kafka-bootstrap:9092
  - alias: "my-cluster-target"
    config:
      offset.flush.timeout.ms: 10000
      producer.buffer.memory: 8388608
    bootstrapServers: my-cluster-target-kafka-bootstrap:9092
  mirrors:
  - sourceCluster: "my-cluster-source"
    targetCluster: "my-cluster-target"
    sourceConnector:
      tasksMax: 10
----

== Checking the message flow

If you are using Prometheus and Grafana to monitor your deployment, you can check the MirrorMaker 2.0 message flow.
The example MirrorMaker 2.0 Grafana dashboard provided with Strimzi shows the following metrics related to the flush pipeline.

> * A count of the Kafka Connect message queue

Reviewer: ```suggestion
* The number of messages in the Connect's outstanding messages queue
```
